### PR TITLE
Multistage behavior + vertex count + minmax dist

### PR DESCRIPTION
**Pull request by Jrius**
Adds support for Multistage Behaviors in AlcScript (see http://forum.guildofwriters.org/viewtopic.php?f=9&t=6365).
Bumps the maximum vertex count per object to 0xffff (see http://forum.guildofwriters.org/viewtopic.php?p=52159#p52159).
Adds support for "mindist" and "maxdist" as object properties. These are bound to the icicle's fMinDist and fMaxDist (not really useful but didn't bother removing it).

### CODE REVIEW

**Comment by Deledrius** [comment]
Thanks for contributing this!  I'll try to look through it all as soon as I can.

File: PyPRP/prp_AnimClasses.py [hunk 1]
```diff
@@ -672,6 +672,50 @@ def write(self, s):
...
+                elif flag == "advance": animstage.fNotify |= plAnimStage.NotifyType["kNotifyAdvance"]
+                elif flag == "loop": animstage.fNotify |= plAnimStage.NotifyType["kNotifyLoop"]
+                elif flag == "regress": animstage.fNotify |= plAnimStage.NotifyType["kNotifyRegress"]
+        
+        animstage.fForwardType    = plAnimStage.ForwardType[     "kForward" + str(FindInDict(argscript,"forward", "none")).capitalize()  ]
+        animstage.fBackType       = plAnimStage.BackType[        "kBack" + str(FindInDict(argscript,"backward", "none")).capitalize()  ]
+        animstage.fAdvanceType    = plAnimStage.AdvanceType[     "kAdvance" + str(FindInDict(argscript,"advance", "none")).capitalize()  ]
+        animstage.fRegressType    = plAnimStage.RegressType[     "kRegress" + str(FindInDict(argscript,"regress", "none")).capitalize()  ]
+        
+        advto = int(FindInDict(argscript,"advanceto",-3125)) # if the line is not present, it will read as -3125,
```
Comment (Deledrius) [inline PyPRP/prp_AnimClasses.py]: Where does this number come from?
I don't see this magic number in plMultiStage.
According to [plMultistageStage.h](https://github.com/H-uru/Plasma/blob/master/Sources/Tools/MaxComponent/plMultistageStage.h#L108) it's a `uint32`, and I see no similar special casing in the [reading or writing routines](https://github.com/H-uru/Plasma/blob/master/Sources/Tools/MaxComponent/plMultistageStage.cpp#L135).
The same is true for the comparable parts of libHSPlasma's [plAnimStage.h](https://github.com/H-uru/libhsplasma/blob/master/core/PRP/Avatar/plAnimStage.h)/[plAnimStage.cpp](https://github.com/H-uru/libhsplasma/blob/master/core/PRP/Avatar/plAnimStage.cpp)

**Comment by Jrius** [comment]
Thank you for looking so quickly into this !
Yeah, there are some bad coding habits I should get rid of... I'm not really used to how Github works either. Plus, I originally made the changes on an outdated version of PyPRP which was not synced with Git... not the best of choice ;)
I'll change all this according to your suggestions.

**Comment by Hoikas** [comment]
Changing the maximum vertex count to 0xFFFF introduces crash bugs into ages that are converted via PlasmaTransfusion and run by CWE. This is probably a CWE bug, but it was easier to fix it here.

File: PyPRP/prp_AnimClasses.py [hunk 2]
```diff
@@ -672,6 +672,50 @@ def write(self, s):
...
+regress: <none|auto|onMove|onAnyKey|max>
+advanceto: <stage number, ex: 0 (first stage)> # optional, if not present: is disabled entirely
+regressto: <stage number, ex: 3 (fourth stage)> # optional, if not present: is disabled entirely
+"""
+        animstage.fAnimName = str(FindInDict(argscript,"animation", "Dance"))
+        animstage.fLoops = int(FindInDict(argscript,"loops", 0))
+        animstage.fNotify = 0
+        notifyflags = FindInDict(argscript,"flagsnotif",[])
+        for flag in notifyflags:
+            if type(flag) == str:
```
Comment (Hoikas) [inline PyPRP/prp_AnimClasses.py]: `if isinstance(flag, str):`

File: PyPRP/prp_AnimClasses.py [hunk 3]
```diff
@@ -672,6 +672,50 @@ def write(self, s):
...
+                if flag == "enter": animstage.fNotify |= plAnimStage.NotifyType["kNotifyEnter"]
+                elif flag == "advance": animstage.fNotify |= plAnimStage.NotifyType["kNotifyAdvance"]
+                elif flag == "loop": animstage.fNotify |= plAnimStage.NotifyType["kNotifyLoop"]
+                elif flag == "regress": animstage.fNotify |= plAnimStage.NotifyType["kNotifyRegress"]
+        
+        animstage.fForwardType    = plAnimStage.ForwardType[     "kForward" + str(FindInDict(argscript,"forward", "none")).capitalize()  ]
```
Comment (Hoikas) [inline PyPRP/prp_AnimClasses.py]: unpythonic spacing

File: PyPRP/prp_AnimClasses.py (hunk 1 above)
Comment (Hoikas) [inline PyPRP/prp_AnimClasses.py]: It would be easier to let the default be `None` and fill in appropriately. There's too much of this unlabelled magic number in the code below.

File: PyPRP/prp_DrawClasses.py [hunk 4]
```diff
@@ -1375,6 +1375,13 @@ def export_obj(self,obj,dynamic,MaterialGroups=[],water = False):
                 # the old way
                 print "Warning: setting icicle properties as int is depreciated"
                 icicle.fProps = flags
+            
+            
+            
+            ######### Sirius' parameters for building LoDs BEGIN #############
+            icicle.fMinDist = getStrIntPropertyOrDefault(obj,"mindist",-1)
+            icicle.fMaxDist = getStrIntPropertyOrDefault(obj,"maxdist",-1)
```
Comment (Hoikas) [inline PyPRP/prp_DrawClasses.py]: ?

File: PyPRP/prp_GeomClasses.py [hunk 5]
```diff
@@ -769,7 +769,7 @@ def write(self,buf):
         buf.Write16(self.fMaxeBoneIdx)
         buf.Write16(self.fPenBoneIdx)
         buf.WriteFloat(self.fMinDist)
-        buf.WriteFloat(self.fMinDist)
+        buf.WriteFloat(self.fMaxDist) ## Sirius: corrected incorrect value
```
Comment (Hoikas) [inline PyPRP/prp_GeomClasses.py]: no need for a comment. make a separate commit that has a detailed message about the change.

File: PyPRP/prp_LogicClasses.py [hunk 6]
```diff
@@ -1431,6 +1439,78 @@ def write(self, s):
...
+            - <receivertype>: <objectname>
+
+"""
+        print "   [MultistageBehMod %s]"%(str(self.Key.name))
+        # allow object to be customized given a given script....
+
+
+        # check where we should add ourself..
+        remote = FindInDict(script,"remote",None)
+        if not remote is None and str(remote) != "":
```
Comment (Hoikas) [inline PyPRP/prp_LogicClasses.py]: `if remote is not None and str:`

File: PyPRP/prp_LogicClasses.py [hunk 7]
```diff
@@ -1431,6 +1439,78 @@ def write(self, s):
...
+
+
+        self.fFreezePhys = bool( str(FindInDict(script,'freezephys'   ,str(self.fFreezePhys)   ) ).lower() == 'true')
+        self.fSmartSeek = bool( str(FindInDict(script,'smartseek'   ,str(self.fFreezePhys)   ) ).lower() == 'true')
+        self.fReverseFBControlsOnRelease = bool( str(FindInDict(script,'reverseonrelease'   ,str(self.fFreezePhys)   ) ).lower() == 'true')
+        
+
+        stagelist = list(FindInDict(script,"stages",[]))
+        for stagescript in stagelist:
+            if type(stagescript) == dict:
```
Comment (Hoikas) [inline PyPRP/prp_LogicClasses.py]: `if isinstance(stagescript, dict):`

**Comment by Deledrius** [comment]
Looks good to me.  The updated diff includes all of the requested fixes.  I'll give this a test and see how the output is.

**Comment by Hoikas** [comment]
Does it replace all the hacks we needed for the GoMePub?

**Comment by Deledrius** [comment]
Yup, this replaces the hacks and exports correctly (with a little adjustment for the better AlcScript). 

:+1:
